### PR TITLE
add numpy 2 hotfix to main

### PR DESCRIPTION
**Pull request by JamesRobertsonGames**
This PR introduces new functionality for checking and updating package dependencies for numpy 2.0 compatibility. It also includes significant updates to the documentation to reflect these changes and provide guidance on using the new features.
### Links

- [PKG-4837](https://anaconda.atlassian.net/browse/PKG-4837) 

### Explanation of changes:
- Added new script `numpy2.py` for checking and proposing numpy 2.0 compatibility updates
- Modified `main.py` to incorporate changes from `proposed_numpy_changes.json`
- Updated README with instructions for using `numpy2.py` and the updated `main.py`
- Added functionality to generate CSV files for reviewing proposed changes
- Included guidance in README for reviewing CSV updates using various tools
- Refactored existing code for better readability and compliance with flake8 standards
- Added error handling and logging improvements in the scripts
- Updated documentation to reflect the new workflow involving `numpy2.py` and `main.py`

[PKG-4837]: https://anaconda.atlassian.net/browse/PKG-4837?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Note to Reviewers:**

This implementation has been carefully designed. However, I value your opinions and welcome any feedback on the design or implementation. This is the best attempt currently at a balanced approach considering future hotfix applications will need to be done carefully. 

Also the packages in the "protect_dict" are just placeholder for now, once this is approved as an approach there will be work to refine this list

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -261,6 +261,42 @@
...
+            if '==' not in dep and '<' not in dep:
+                depends[i] = dep + ",<2.0a0"
+                updated = True
+            elif dep == "numpy":
+                depends[i] = "numpy <2.0a0"
+                updated = True
+            # Add an upper bound if it is missing in the dependency constraint
+            elif '==' not in dep and '<' not in dep:
+                depends[i] = dep + ",<2.0a0"
+                updated = True
```
Comment (cbouss) [inline main.py:294]: I see in the output lines like "numpy-base 1.11.3 py27hdbf6ddf_8", replaced by "numpy-base 1.11.3 py27hdbf6ddf_8,<2.0a0",. The upper bound is added on the hash, which I expect to cause trouble, but most of all, we don't need to add upper bounds to packages pinned to a specific version.

File: main.py [hunk 2]
```diff
@@ -261,6 +261,42 @@
...
+    if updated:
+        instructions["packages"][fn]['depends'] = depends
+    return updated
+
+
+def _update_numpy_dependencies(fn, record, instructions):
+    depends = record.get("depends", [])
+    updated = False
+    for i, dep in enumerate(depends):
+        if dep.startswith("numpy"):
```
Comment (cbouss) [inline main.py:283]:
```suggestion
        if dep.split()[0] == "numpy":
```
Otherwise it will also match numpy-base, numpydoc...

File: main.py [hunk 3]
```diff
@@ -261,6 +261,42 @@
 ]
 
 
+def _update_numpy_base_dependencies(fn, record, instructions):
+    depends = record.get("depends", [])
+    updated = False
+    for i, dep in enumerate(depends):
+        if dep.startswith("numpy-base") and '==' not in dep and '<' not in dep:
```
Comment (cbouss) [inline main.py:268]:
```suggestion
        if dep.split()[0] == "numpy-base" and '==' not in dep and '<' not in dep:
```

File: main.py [hunk 4]
```diff
@@ -261,6 +261,42 @@
...
+            depends[i] = dep + ",<2.0a0"
+            updated = True
+        elif dep == "numpy-base":
```
Comment (cbouss) [inline main.py:271]:
```suggestion
        else:
```

**Comment by JamesRobertsonGames** [comment]
@cbouss thanks for the review, made the suggestions and reuploaded the most upto date output

File: main.py [hunk 5]
```diff
@@ -261,6 +261,38 @@
...
+def _update_numpy_base_dependencies(fn, record, instructions):
+    depends = record.get("depends", [])
+    updated = False
+    for i, dep in enumerate(depends):
+        if dep.split()[0] == "numpy-base" and '==' not in dep and "1." not in dep and '<' not in dep:
+            depends[i] = dep + ",<2.0a0"
+            updated = True
+        else:
+            depends[i] = "numpy-base <2.0a0"
+            updated = True
```
Comment (cbouss) [inline main.py:273]: This causes all dependencies to be replaced by numpy-base

**Comment by JamesRobertsonGames** [comment]
Much better output now in the output.txt @cbouss

**Comment by cbouss** [comment]
Besides the diff it would be great to have some analysis on the repodata before and after the patch. I feel this would help ensure all is covered.

Also, I have been thinking, what if we publish this patch, then start building python packages with no upper bounds (because they support numpy 2), and someone regenerate a patch? A <1 bound would be wrongfully added. So perhaps instead we could have a configuration file like:
```
pandas: 2.2.2,
scikit-lear: 1.4.2,
...
```
Which could be used to patch like (pseudo code):
```
if pkg in numpy2_protect_dict:
  if VersionOrder(version) <= VersionOrder(numpy2_protect_dict['version']):
    add_dep(numpy < 2.0a0)
```
This way if needed we can tweak the configuration file and retain more control.

File: main.py [hunk 6]
```diff
@@ -260,6 +261,60 @@
     "_anaconda_depends",
 ]
 
+# Load the configuration file
+with open('numpy2_protect.yaml', 'r') as f:
+    numpy2_protect_dict = yaml.safe_load(f)
```
[the remark this answers is not in the document]
Comment (JamesRobertsonGames) [reply main.py:266]: My only concern is it is likely to be fairly long? I could add it in a new python file (which just has a giant dict) and include that? This `main.py` is already bloated and long I don't really want to add so 100's of lines potentially (it's 3 at the moment but we haven't compiled the final list)

[the remark this answers is not in the document]
Comment (Jrice1317) [reply main.py:266]: I agree adding pyyaml does seem a bit like overkill for this single usage. Perhaps since `json` is already being imported, that can be utilized here?

[the remark this answers is not in the document]
Comment (JamesRobertsonGames) [reply main.py:266]: `json` wasn't selected just because it's slightly less human readable, might as well go forward with a dict in a separate file I think that strikes the balance, and if I do end up writing a script to generate it I can still do that with a separate .py file

File: README.md [hunk 7]
```diff
@@ -45,12 +45,76 @@ We should put things on the revoke list when:
...
+python numpy2.py
+```
+
+### What numpy2.py does
+
+`numpy2.py` performs the following tasks:
+1. Scans through the repodata for packages depending on numpy.
+2. Checks if these dependencies need updates to ensure compatibility with numpy 2.0.
+3. Proposes changes to add upper bounds to numpy dependencies where necessary.
+4. Generates a `proposed_numpy_changes.json` file containing all proposed changes.
```
Comment (cbouss) [inline README.md:64]: Can you rename `proposed_numpy_changes.json` to `numpy2_patch.json`? Since this file is checked in and will remain this will be clearer.

File: main.py [hunk 8]
```diff
@@ -261,6 +275,134 @@
 ]
 
 
+def load_numpy_changes():
+    try:
+        with open('proposed_numpy_changes.json', 'r') as f:
```
Comment (cbouss) [inline main.py:280]: Not having this file should fail the script.

File: main.py [hunk 9]
```diff
@@ -460,6 +602,20 @@ def _fix_cudnn_depends(depends, subdir):
...
+
+    # Loop through each key-value pair in the 'Packages' dictionary
+    for key, value in repo_data.items():
+        # If the key contains the desired substring, add it to the filtered dictionary
+        if substring in key:
+            filtered_dict[key] = value
+
+    return filtered_dict
+
+
```
Comment (cbouss) [inline main.py:618]:
```suggestion
```
I don't see that function being used.

File: main.py [hunk 10]
```diff
@@ -261,6 +275,134 @@
...
+            return json.load(f)
+    except FileNotFoundError:
+        logger.warning("proposed_numpy_changes.json not found. No numpy changes will be applied.")
+        return {}
+
+
+NUMPY_CHANGES = load_numpy_changes()
+
+
+def apply_numpy_changes(record, subdir, filename):
```
Comment (cbouss) [inline main.py:290]: Can you make it more explicit? Perhaps just replacing numpy by numpy2.

File: numpy2.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    "linux-64",
+    "linux-aarch64",
+    "linux-armv6l",
+    "linux-armv7l",
+    "linux-ppc64le",
+    "linux-s390x",
+    "osx-64",
+    "osx-arm64",
+    "win-32",
+    "win-64",
```
Comment (cbouss) [inline numpy2.py:39]:
```suggestion
    "noarch",
    "linux-64",
    "linux-aarch64",
    "linux-s390x",
    "osx-64",
    "osx-arm64",
    "win-64",
```
Only need to work on what we support today.

File: numpy2.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+                )
+    for subdir in SUBDIRS:
+        index = repodatas[subdir]["packages"]
+        for fn, record in index.items():
+            name = record["name"]
+            depends = record["depends"]
+            constrains = record.get("constrains", [])
+
+            depends = [dep for dep in depends if dep is not None]
+            if "py39" in fn or "py310" in fn or "py311" in fn or "py312" in fn:
```
Comment (cbouss) [inline numpy2.py:234]:
```suggestion
            # numpy 2 is introduced for python 3.9. Packages for python 3.13 will be built with numpy2 from the start.
            if "py39" in fn or "py310" in fn or "py311" in fn or "py312" in fn:
```

Comment (cbouss) [reply numpy2.py:234]: This leaves noarch packages out though. Perhaps that's fine.

**Review by cbouss — COMMENTED** [review]
Did a quick first pass

File: numpy2_patch.json
Comment (cbouss) [inline numpy2_patch.json:1]: Can you generate it avoiding the case where no update actually occurs?
I mean removing entries like:
```
    "adtk-0.6.2-py310h2f386ee_0.tar.bz2": [
      {
        "type": "dep",
        "original": "numpy >=1.21.5,<2.0a0",
        "updated": "numpy >=1.21.5,<2.0a0",
        "reason": "Already has upper bound"
      }
    ],
```

Comment (JamesRobertsonGames) [reply numpy2_patch.json:1]: Done and regenerated the file 👍

File: main.py [hunk 13]
```diff
@@ -273,6 +284,95 @@
...
+    Applies predefined numpy changes to a record based on its directory and filename.
+
+    Parameters:
+    - record: The record to update.
+    - subdir: The subdirectory of the record.
+    - filename: The filename of the record.
+    """
+    if subdir not in NUMPY_2_CHANGES or filename not in NUMPY_2_CHANGES[subdir]:
+        return
+    changes = NUMPY_2_CHANGES[subdir][filename]
```
Comment (ryanskeith) [inline main.py:310]: You can then refer to subdir_changes from above.

File: main.py [hunk 14]
```diff
@@ -273,6 +284,95 @@
...
+    - list: The list of dependencies or constraints based on the change type, None if the change type is unrecognized.
+    """
+    if change_type == 'dep':
+        return record['depends']
+    elif change_type == 'constr':
+        return record.get('constrains', [])
+    return None
+
+
+def _apply_changes_to_dependencies(depends, change, record, filename, sort_type='reason'):
```
Comment (ryanskeith) [inline main.py:336]:
```suggestion
def _apply_changes_to_dependencies(depends_or_constraints, change, record, filename, sort_type='reason'):
```

**Review by ryanskeith — CHANGES_REQUESTED** [review]
After thinking on this over the weekend, I feel there are two major things to ask for (if I understand the code correctly).

First, let's simplify numpy2.py code.
- Let's rename to generate_numpy2.py changes
- Remove all code not being used and most reporting type code.
- Simplify the data structure in the numpy2_patch.json
- Move testing code to the tests folder

Second, let's simplify the incorporation of the numpy2 patches into main.py
- Remove any logging or side product reporting
- Use standard dep replacement code

File: numpy2.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    - dependencies_list: Dependencies to check and modify.
+    - package_record: Metadata about the current package.
+    - dependency_type: Type of dependency ('run', 'build').
+    - package_subdir: Package location subdirectory.
+    - filename: Package filename.
+    """
+    for i, dependency in enumerate(dependencies_list):
+        parts = dependency.split()
+        package_name = parts[0]
+        if package_name in ["numpy", "numpy-base"]:
```
[the remark this answers is not in the document]
Comment (JamesRobertsonGames) [reply numpy2.py:155]: As discussed: this is correct right now as we have not properly populated the protected versions

**Comment by JamesRobertsonGames** [comment]
@cbouss @ryanskeith 

Should have addressed the comments, the output is looking good on the test-hotfix.py to me, removed all non-essential logging and simplified the solution based on the suggestions. Let me know if I have missed anything or anything new is observed :)

File: main.py [hunk 16]
```diff
@@ -273,6 +273,51 @@
...
+        return
+
+    for change in relevant_changes:
+        depends = _get_dependency_list(record, change['type'])
+        if depends is None:
+            continue
+        replace_dep(depends, change["original"], change["updated"])
+
+
+def _get_dependency_list(record, change_type):
```
Comment (ryanskeith) [inline main.py:303]: I don't think this warrants its own function.  If you store "depends" and "constrains" directly in the numpy2_patch.json file, the replace_dep line becomes:

```
replace_dep(record[change["change_type"]], change["original"], change["updated"]
```

Your process already did all the cleaning so we shouldn't have any weird third cases.

Comment (ryanskeith) [reply main.py:303]: This also gets rid of the unnecessary for loop.

File: main.py [hunk 17]
```diff
@@ -686,6 +731,9 @@ def patch_record_in_place(fn, record, subdir):
                 depends[i] = depends[i].replace(">=1.21.5,", ">=1.21.2,")
                 break
 
+    if NUMPY_2_CHANGES:
```
Comment (ryanskeith) [inline main.py:734]: As stated before, as this block changes many different files and NOT a targeted file.  It should be run last.

File: generate_numpy2_patch.py [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+    for _, dependency in enumerate(dependencies_list):
+        parts = dependency.split()
+        package_name = parts[0]
+        if package_name in ["numpy", "numpy-base"]:
+            if not has_upper_bound(dependency):
+                if package_name in numpy2_protect_dict:
+                    version_str = parts[1] if len(parts) > 1 else None
+                    version = parse_version(version_str) if version_str else None
+                    protected_version = parse_version(numpy2_protect_dict[package_name])
+                    if version and protected_version:
```
[the remark this answers is not in the document]
Comment (ryanskeith) [reply generate_numpy2_patch.py:141]: ah, I see.

File: main.py [hunk 19]
```diff
@@ -273,6 +273,30 @@
...
+    - filename: The filename of the record.
+    """
+    if subdir not in NUMPY_2_CHANGES:
+        return
+
+    change = NUMPY_2_CHANGES[subdir].get(filename)
+    if not change:
+        return
+    else:
+        try:
```
Comment (ryanskeith) [inline main.py:294]: In what case do you feel that you will have a failure when you have been given a filename the record will fail?  I don't think you need to be so protective here.

File: generate_numpy2_patch.py [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+            depends = record["depends"]
+            constrains = record.get("constrains", [])
+
+            depends = [dep for dep in depends if dep is not None]
+            if any(py_ver in fn for py_ver in ["py39", "py310", "py311", "py312"]):
+                if name not in ["anaconda", "_anaconda_depends", "__anaconda_core_depends", "_anaconda_core"]:
+                    try:
+                        for dep in depends:
+                            if dep.split()[0] in ["numpy", "numpy-base"]:
+                                update_numpy_dependencies(depends, record, "dep", subdir, fn)
```
Comment (ryanskeith) [inline generate_numpy2_patch.py:198]:
```suggestion
                                update_numpy_dependencies(depends, record, "depends", subdir, fn)
```

File: generate_numpy2_patch.py [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+                        for constrain in constrains:
+                            if constrain.split()[0] in ["numpy", "numpy-base"]:
+                                update_numpy_dependencies(constrains, record, "constr", subdir, fn)
```
Comment (ryanskeith) [inline generate_numpy2_patch.py:201]:
```suggestion
                                update_numpy_dependencies(constrains, record, "constrains", subdir, fn)
```

File: generate_numpy2_patch.py [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+    Parameters:
+    - subdirectory: The subdirectory where the file is located.
+    - filename: The name of the file being modified.
+    - change_type: The type of change (e.g., 'dep', 'constr').
+    - original_dependency: The original dependency string.
+    - updated_dependency: The updated dependency string.
+    - reason: The reason for the change.
+    """
+    # change dep and constr to dependency and constraint
+    if change_type == "dep":
```
Comment (ryanskeith) [inline generate_numpy2_patch.py:55]: By giving the full change_type on creation, you don't need this if else statement.

File: main.py [hunk 23]
```diff
@@ -1444,6 +1466,14 @@ def patch_record_in_place(fn, record, subdir):
         if name == "clangxx_osx-64" and version == "4.0.1" and int(build_number) < 17:
             depends[:] = ["clang_osx-64 >=4.0.1,<4.0.2.0a0", "clangxx", "libcxx"]
 
+    ###########
+    # numpy 2 #
+    ###########
+
+    # Numpy 2.0.0 Patch
```
Comment (ryanskeith) [inline main.py:1473]: Needs comments on what this does.

File: main.py [hunk 24]
```diff
@@ -273,6 +273,30 @@
...
+    - subdir: The subdirectory of the record.
+    - filename: The filename of the record.
+    """
+    if subdir not in NUMPY_2_CHANGES:
+        return
+
+    change = NUMPY_2_CHANGES[subdir].get(filename)
+    if not change:
+        return
+    else:
```
Comment (ryanskeith) [inline main.py:293]: We don't need an else in this case.

File: generate_numpy2_patch.py [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+                            if dep.split()[0] in ["numpy", "numpy-base"]:
+                                update_numpy_dependencies(depends, record, "dep", subdir, fn)
+                        for constrain in constrains:
+                            if constrain.split()[0] in ["numpy", "numpy-base"]:
+                                update_numpy_dependencies(constrains, record, "constr", subdir, fn)
+                    except Exception as e:
+                        logger.error(f"numpy 2.0.0 error {fn}: {e}")
+
+    json_filename = Path("numpy2_patch.json")
+    with json_filename.open('w') as f:
```
Comment (ryanskeith) [inline generate_numpy2_patch.py:206]: Not necessary but this can be a one liner.
```suggestion
    json_filename.write_text(json.dumps(dict(NUMPY_2_CHANGES), indent=2)
```

**Review by ryanskeith — CHANGES_REQUESTED** [review]
This is progressing quite well.  A couple things.

- I think we can safely drop "numpy-base" from the list as it has never been triggered
  - This should simplify the logic in places so you can directly ask and not have to do the reverse.
- I think as this code has a lot of nested if/else statements, you need more comments inline as the logic is hard to follow.  My gut tells me there is still some simplifications that should occur.
  - Specifically more comments for generate_numpy2_patch:update_numpy_dependencies and :main
- Lastly, how are you checking for no arch again?  the filename will not work in the noarch case.

**Comment by JamesRobertsonGames** [comment]
@ryanskeith done your suggestions, on the noarch I think a previous review concluded that there is little point doing noarch in this instance? @cbouss if you have an opinion on this feel free to wade in! Happy to handle this edge case, or we could deploy without this and then rectify the noarch packages in another PR?

**Comment by cbouss** [comment]
> @ryanskeith done your suggestions, on the noarch I think a previous review concluded that there is little point doing noarch in this instance? @cbouss if you have an opinion on this feel free to wade in! Happy to handle this edge case, or we could deploy without this and then rectify the noarch packages in another PR?

Yes I believe it is best to not patch noarch packages, at least for now. I am not sure of which edge cases this would uncover, and it is now time to release numpy 2.

**Review by cbouss — APPROVED** [review]
LGTM!

File: generate_numpy2_patch.py (hunk 22 above)
Comment (ryanskeith) [reply generate_numpy2_patch.py:55]: You can get rid of lines 55-58 now.

**Review by ryanskeith — APPROVED** [review]
Looks good.  You could remove the dead code in `generate_numpy2_patch.py`, but it isn't necessary.
